### ISSUE TITLE
Remove local fork of Golang's archive/tar

### ISSUE THREAD

**Issue by thaJeztah** [moby/moby#42402]
commit aa6a9891b09cce3d9004121294301a30d45d998d (19.03 branch) and https://github.com/moby/moby/pull/40672 (master / 20.10) re-introduced a local copy of go's archive/tar package, with a patch applied [patches/0001-archive-tar-do-not-populate-user-group-names.patch](https://github.com/moby/moby/blob/841c1f338872ca37d9fe8c699336823222dba5da/patches/0001-archive-tar-do-not-populate-user-group-names.patch).

This patch was applied for the [19.03.8 release](https://github.com/docker/docker-ce/releases/tag/v19.03.8) to improve mitigation for [CVE-2019-14271](https://nvd.nist.gov/vuln/detail/CVE-2019-14271) for some nscd configuration.

We should try to get rid of this fork again.

The discussion on https://github.com/moby/moby/pull/40672#discussion_r391306456 mentioned we should open a ticket / pull request in upstream Go to make this functionality "optional", but I think @tonistiigi also had alternatives in mind to address it.

**Comment by tianon** [comment]
I'm definitely all for it, especially since in Go 1.17+ we're going to have a _much_ harder time maintaining this the way we currently do (and I think it already probably causes grief for packagers who don't get to use our `vendor/` as-is).

@cpuguy83 suggested that if we still need the patch we could instead maintain a proper forked package, which would be at least possible to deal with in a Go modules world, but IMO something that doesn't require forking the standard library would be awesome if there are ideas around that :sweat_smile:

**Comment by errordeveloper** [comment]
We need to revisit in the light of switching away from `vndr` to `go mod vendor` (that's without adding `go.mod` for now).

> ...I think @tonistiigi also had alternatives in mind to address it

 @tonistiigi could you please share what you were thinking in that regard?

**Comment by errordeveloper** [comment]
> @cpuguy83 suggested that if we still need the patch we could instead maintain a proper forked package, which would be at least possible to deal with in a Go modules world...

Are we only concerned with Docker's usage of `archive/tar`, or perhaps there are some dependencies that are affected also?

**Comment by thaJeztah** [comment]
> we could instead maintain a proper forked package, which would be at least possible to deal with in a Go modules world...

I would be ok with doing this for the short term (change the import paths to a package in moby (or could be a separate repo). What I wasn't sure about for this though would be if we're _able_ to replace it everywhere where we want it to be replaced; if `archive/tar` is used indirectly in Go stdlib, is there a risk of reintroducing the issue we're trying to fix?

I haven't looked deeply into that scenario, but possibly others know.

The "proper" fix of course is to propose a patch upstream to add a build-tag to disable the bits we're removing (`nouserlookup` or whatever), but likely go 1.19 would be the first chance to get that in

**Comment by errordeveloper** [comment]
On Fri, 10 Dec 2021, 4:12 pm Sebastiaan van Stijn, ***@***.***>
wrote:

> we could instead maintain a proper forked package, which would be at least
> possible to deal with in a Go modules world...
>
> I would be ok with doing this for the short term (change the import paths
> to a package in moby (or could be a separate repo). What I wasn't sure
> about for this though would be if we're *able* to replace it everywhere
> where we want it to be replaced; if archive/tar is used indirectly in Go
> stdlib, is there a risk of reintroducing the issue we're trying to fix?
>
Do we know what code in Moby is calling stdlib to pack/unpack tarballs and
introduces problematic behaviour?

>

**Comment by tonistiigi** [comment]
> @tonistiigi could you please share what you were thinking in that regard?

I don't really remember anymore. I think @cpuguy83 had a suggestion about separating out this code to another binary and guaranteeing that it is always built without cgo(or with musl). I don't think it is very safe though, especially easy for some 3rd party build to mess it up.

> Do we know what code in Moby is calling stdlib to pack/unpack tarballs and
introduces problematic behaviour?

IIRC, the extract code is doing a reexec of dockerd with a chroot applied to protect against breakouts. If that code now running inside chroot is built with cgo and upstream tar package, when it will call `lookupuser` it will call the equivalent glibc method. That glibc method will at runtime try to map some `.so` files to executable memory and run it. Inside chroot that is of course problematic as it would load the `.so` that is inside the attacker controlled extract path, not the one from the host system.

**Comment by errordeveloper** [comment]
On Fri, 10 Dec 2021, 6:20 pm Tõnis Tiigi, ***@***.***> wrote:

> @tonistiigi <https://github.com/tonistiigi> could you please share what
> you were thinking in that regard?
>
> I don't really remember anymore. I think @cpuguy83
> <https://github.com/cpuguy83> had a suggestion about separating out this
> code to another binary and guaranteeing that it is always built without
> cgo(or with musl). I don't think it is very safe though, especially easy
> for some 3rd party build to mess it up.
>
Maybe a runtime check could just panic to avoid this? Just an idea...

> Do we know what code in Moby is calling stdlib to pack/unpack tarballs and
> introduces problematic behaviour?
>
> IIRC, the extract code is doing a reexec of dockerd with a chroot applied
> to protect against breakouts. If that code now running inside chroot is
> built with cgo and upstream tar package, when it will call lookupuser it
> will call the equivalent glibc method. That glibc method will at runtime
> try to map some .so files to executable memory and run it. Inside chroot
> that is of course problematic as it would load the .so that is inside the
> attacker controlled extract path, not the one from the host system..
>
It does sound like we should be in control of this, I'd surprised if the
tarball extraction is done indirectly (and even if happens to be, surely it
can addressed).
Also, for extra safety a runtime check could still be of interest, perhaps.
This definitely does need a test case.

**Comment by kolyshkin** [comment]
Better late than never: https://go-review.googlesource.com/c/go/+/371054

The alternative approach is to add something like `archivetarnonames` (to disable name lookups), but since these lookups are only done by a single function that is not used internally, it seems better to add another.

**Comment by kolyshkin** [comment]
For the lack of better things to do on a Friday night, wrote a proposal: https://github.com/golang/go/issues/50102. Most probably it has errors, omissions and other issues -- please review and suggest improvements.

**Comment by thaJeztah** [comment]
❤️ @kolyshkin !

**Comment by errordeveloper** [comment]
Thanks @kolyshkin! I added a few comments, hopefully will be somewhat helpful :)

**Comment by corhere** [comment]
I took a look at the situation for myself and have some possible ideas for how to get rid of the fork without regressing on the vulnerability or waiting for changes to be accepted into the upstream standard library.

The root cause of the vulnerability, AIUI, is that the cgo/POSIX implementation of `os/user` calls libc functions which, in glibc, are implemented by dynamically loading an NSS module (`libnss_*.so.2`) into the process. An attacker could trick the process into loading a malicious attacker-controlled library by setting `LD_LIBRARY_PATH` in the process's environment or dropping the malicious library into one of the default library search paths. This is not normally a security vulnerability as an attacker would need to already have enough privilege on the host that there would be no need to exploit the process: if they already have write access to one of the host's default library paths or the ability to control the `LD_LIBRARY_PATH` environment variable, they would already have enough access to e.g. replace `libc.so.6`, set `LD_PRELOAD` on the process to inject their malicious code directly, or use their (most likely) root privileges to simply replace the `dockerd` binary outright.

Ironically, the hardening of moby's archiver using a chroot is what makes the dynamic loading of NSS modules into an exploitable attack vector. The attacker has full control of the entire chroot filesystem — it's a container's filesystem! If the first call to one of the `os/user` functions which triggers the loading of the NSS modules happens after pivoting into the chroot, the modules are resolved and loaded _from the attacker-controlled container filesystem._ There was an attempt to mitigate the vulnerability with #39612 which ensures that the NSS modules are loaded into the process before pivoting into the chroot. But seeing as the forking of `archive/tar` was re-added after, I can only assume that loading the host's NSS modules then pivoting into what is effectively a completely different OS userland could still break things. So the only complete fix would be to avoid calling any code which calls into the libc functions that trigger glibc to load NSS modules into the process. Building the binary with `CGO_ENABLED=0` or the `osusergo` build flag is sufficient to ensure that none of the `os/user` functions call into libc, but that has some downsides which @kolyshkin has already elaborated upon in his proposal on golang/go. Without those build configuration changes, the issues can be avoided "merely" by ensuring that the `os/user` functions are never called by the chroot'ed process. This is easier said than done.

The function `archive/tar.FileInfoHeader` calls the problematic `os/user` functions via the unexported function `tar.statUnix`, indirectly (more on that in a moment). `FileInfoHeader` is the only function which calls `statUnix`. The `FileInfoHeader` function is used by moby when creating (not extracting) archives within the chroot. And it's a leaf function: no code in Go's standard library (1.16 and 1.17) calls that function outside of tests. `FileInfoHeader` returns a `tar.Header` struct, which has no unexported members, therefore a `tar.Header` value could be fully constructed from normal application/library code. _The `FileInfoHeader` function is not strictly required._ Therefore, for application code fully under our control, an alternative implementation which doesn't call `os/user` functions can be subbed in without any trickery, patches or hacks. This could take the form of forking and patching a copy of the standard library's implementation of `tar.FileInfoHeader` (e.g. `github.com/docker/docker/internal/tarhdr`) [or writing a totally new implementation that does exactly what moby needs it to do.](https://github.com/moby/moby/blob/79651b7a979b40e26af353ad283ca7ea5d67a855/pkg/archive/archive.go#L406-L425) This would Just Work... so long as there isn't any third-party code which indirectly calls `tar.FileInfoHeader` unexpectedly. Since the chroot-archive process is limited enough in scope(?) that can be audited manually, but it would be nice to have defense-in-depth. That's where patching `archive/tar` directly would be helpful. If only there was a way to do it in a Go Modules-compatible way.

The implementations of `tar.FileInfoHeader` in Go 1.16 and 1.17 just so happen to be written in such a way that the call to `tar.statUnix` can be monkeypatched with an alternative implementation at runtime. `statUnix` is called [indirectly through a function variable](https://cs.opensource.google/go/go/+/refs/tags/go1.17.6:src/archive/tar/common.go;l=701-703;bpv=1;bpt=1) which is [assigned at runtime](https://cs.opensource.google/go/go/+/refs/tags/go1.17.6:src/archive/tar/stat_unix.go;l=19-21). It's not exported, but there is an (unsafe) way to assign to it from a different package, anyway.
```go
import (
	"archive/tar"
	"io/fs"

	_ "unsafe" // required for go:linkname
)

//go:linkname sysStat archive/tar.sysStat
var sysStat func(fi fs.FileInfo, h *tar.Header) error
```
With this one ~weird trick~ little hack, moby code can assign `sysStat = nil` or `sysStat = myStatUnix` to ensure that any code calling into `tar.FileInfoHeader` anywhere in the process would not invoke the problematic libc functions! It could break between Go versions, but multiple versions of Go could be supported in the same codebase by utilizing build constraints. It is not ideal, but I think it would be a much more maintainable situation than the status quo while Go's standard library catches up.

What do you think? I would be happy to share my proof-of-concept, or open a PR which implements what I have proposed.

**Comment by corhere** [comment]
The big caveat for my ideas above is how brittle the defenses are: any call to an `os/user` function after the process has pivoted into the chroot could reintroduce the vulnerability. Extreme vigilance and static analysis would be required. Ideally we could robustly harden the chroot-archiver such that glibc `dlopen()`ing one of its modules while the process is chrooted cannot lead to execution of malicious code. I have some ideas on how to do just that.

[The dynamic linker](https://man7.org/linux/man-pages/man8/ld.so.8.html) and [its runtime counterpart `dlopen()`](https://man7.org/linux/man-pages/man3/dlopen.3.html) searches for the shared object to load by checking:
1. the directories specified in the `DT_RPATH` or `DT_RUNPATH` ELF tags embedded into the binary,
2. the directories specified in the `LD_LIBRARY_PATH` environment variable,
3. for an entry in the file `/etc/ld.so.cache` specifying the path to the object,
4. in the default paths `/lib{64}` and `/usr/lib{64}`.

I believe that removing the attacker's ability to control the contents of the root directory of the chroot filesystem is sufficient to prevent them from tricking the dynamic linker into loading shared objects from within the chroot. Mounting the container's root filesystem such that it is a subdirectory of the chroot stops the attacker from controlling the contents of the trusted `/etc`, `/usr` and `/lib` trees.

The values of `DT_RPATH`, `DT_RUNPATH` and `LD_LIBRARY_PATH` are not attacker-controlled in our threat model. The `DT_*` ELF tags are set when the binary is built, therefore are controlled by the packagers/distributors of moby. The value of `LD_LIBRARY_PATH` is controlled by the system administrator. We can guard against exploits via these configurations, even by a malicious administrator, by randomizing the directory name for the container filesystem's mountpoint to make it unpredictable.

-----

For more defense-in-depth, the container filesystem can be (bind-)mounted `noexec`. The Linux kernel blocks attempts to `mmap(PROT_EXEC)` memory pages from files on noexec filesystems, so even if an attacker found another way to trick the dynamic linker into loading a shared object from the container filesystem, attempts to map executable sections of the file into the process's memory space would fail. This is not yet sufficient to completely block all exploits, however, as relocations can be abused to build and execute a Return-Oriented Programming chain before the linker attempts to map any executable pages ([sourceware bugzilla](https://sourceware.org/bugzilla/show_bug.cgi?id=21718), [ChromeOS issue](https://bugs.chromium.org/p/chromium/issues/detail?id=1182687)). There is ongoing work on both the kernel and userland sides to plug this hole upstream ([glibc patch](https://sourceware.org/pipermail/libc-alpha/2021-July/129282.html), [kernel support](https://lwn.net/Articles/832959/)) and security-hardened systems such as ChromeOS are already incorporating patches into their distributions. Once the upstream work is complete and is incorporated into shipping systems, we would get the more robust hardening with no further changes to moby code.

-----

By mounting the container filesystem as a randomly-named subdirectory of the chroot with the `noexec` option, we can ensure that bugs in moby code which trigger dynamic linking while chrooted cannot be exploited to escape the sandbox via code execution. It would at worst result in a non-exploitable crash of the chroot-archive process. Therefore (monkey)patching the Go standard library would not be necessary. Our own custom implementation of `TarInfoHeader` would still be required, but it will no longer be the vulnerability linchpin.

**Comment by cpuguy83** [comment]
Bind mounting into a safe location could work... e.g. `/tmp/tarJob-<rand>/rootfs`

Originally issue I think was that we did chroot to the rootfs as a subdir but this was within the main graphdriver dir tree and allowed the container process to use symlinks to gain access to the host.

Another layer could be putting randomized junk values into all the lookup vars. I'm not sure what glibc would do here if the value doesn't exist, though.

**Comment by corhere** [comment]
The more I think about it, the more uneasy I get about `chroot`ing into a hostile rootfs inside a dynamically-linked process after it has started executing. It's quite the fair assumption that the library search paths which were trusted to locate the shared libraries when the executable image was first linked and loaded will continue to be trustworthy for the duration of the process's lifetime. Any version of any shared library could `dlopen` another library as a side effect of practically anything, as an implementation detail, and it would be completely justified to do so. Any mitigations to harden docker-tar against a dynamically-linked library loading code from inside the chroot could be rendered useless without warning by a glibc upgrade, or linking against musl, or bionic. And the mitigations themselves necessarily have to be complex and clever, which are not desirable properties of security-sensitive code.

I'm really warming up to the idea of building `docker-tar` as a binary with cgo always disabled as it takes native libraries completely out of the equation. And to address @tonistiigi's concern about compile-time safety, third parties can be stopped from building it dynamically-linked by adding a `!cgo` build constraint to one or more critical files.
<details>
<summary>Proof of concept</summary>

```console
$ cat main.go
//go:build !cgo
// +build !cgo

package main

import (
	"fmt"
	"os/user"
)

func main() {
	u, err := user.Lookup("docker")
	fmt.Println(u, err)
}
$ go build
package .: build constraints exclude all Go files in [path]
$ CGO_ENABLED=0 go build
$
```
</details>

With cgo out of the way, the most serious consequence of `archive/tar.TarInfoHeader` getting called while the process is chrooted would be a performance regression from repeatedly parsing `/etc/groups` for every file getting archived.

Of course, `docker-tar`, `docker-untar` and `docker-applyLayer` could all be compiled into a single binary so they all get the hardening benefits of static, cgo-less compilation while amortizing the disk usage across all three.
